File: backend/src/modules/wallet/wallet.service.ts
import { PrismaClient, PayoutStatus } from '@prisma/client';
import { AppError } from '../../middleware/error.middleware';
import { ERROR_CODES, HTTP_STATUS } from '../../utils/constants';
import { logger } from '../../utils/logger';
import { env } from '../../config/env';

const prisma = new PrismaClient();

export class WalletService {
  /**
   * Add funds to pending balance
   * Called when payment is captured
   * 
   * CRITICAL: Uses transaction with row-level locking
   */
  async addPendingBalance(providerId: string, amount: number, metadata?: any) {
    return await prisma.$transaction(async (tx) => {
      // Get or create wallet with SELECT FOR UPDATE (row lock)
      let wallet = await tx.providerWallet.findUnique({
        where: { providerId },
      });

      if (!wallet) {
        wallet = await tx.providerWallet.create({
          data: {
            providerId,
            availableBalance: 0,
            pendingBalance: 0,
            lifetimeEarnings: 0,
            totalWithdrawn: 0,
          },
        });
      }

      // Lock wallet row for update
      const lockedWallet = await tx.$queryRaw<any[]>`
        SELECT * FROM provider_wallets 
        WHERE provider_id = ${providerId}
        FOR UPDATE
      `;

      if (!lockedWallet || lockedWallet.length === 0) {
        throw new AppError('Wallet not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
      }

      const balanceBefore = wallet.pendingBalance;
      const balanceAfter = balanceBefore + amount;

      // Update wallet
      const updatedWallet = await tx.providerWallet.update({
        where: { providerId },
        data: {
          pendingBalance: balanceAfter,
          lifetimeEarnings: { increment: amount },
        },
      });

      // Record transaction
      await tx.walletTransaction.create({
        data: {
          walletId: updatedWallet.id,
          userId: providerId,
          type: 'CREDIT',
          amount,
          balanceBefore,
          balanceAfter,
          description: 'Session payment received (pending)',
          referenceType: metadata?.referenceType || 'session',
          referenceId: metadata?.referenceId,
          metadata,
        },
      });

      logger.info('Added pending balance', {
        providerId,
        amount,
        balanceAfter,
      });

      return updatedWallet;
    });
  }

  /**
   * Release pending balance to available balance
   * Called when session is completed
   * 
   * CRITICAL: Uses transaction with row-level locking
   */
  async releasePendingBalance(providerId: string, amount: number, metadata?: any) {
    return await prisma.$transaction(async (tx) => {
      // Lock wallet
      const wallet = await tx.providerWallet.findUnique({
        where: { providerId },
      });

      if (!wallet) {
        throw new AppError('Wallet not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
      }

      // Lock row
      await tx.$queryRaw`
        SELECT * FROM provider_wallets 
        WHERE provider_id = ${providerId}
        FOR UPDATE
      `;

      // Check sufficient pending balance
      if (wallet.pendingBalance < amount) {
        throw new AppError(
          'Insufficient pending balance',
          HTTP_STATUS.BAD_REQUEST,
          ERROR_CODES.INSUFFICIENT_BALANCE
        );
      }

      const pendingBefore = wallet.pendingBalance;
      const availableBefore = wallet.availableBalance;

      // Update wallet
      const updatedWallet = await tx.providerWallet.update({
        where: { providerId },
        data: {
          pendingBalance: { decrement: amount },
          availableBalance: { increment: amount },
        },
      });

      // Record transaction
      await tx.walletTransaction.create({
        data: {
          walletId: updatedWallet.id,
          userId: providerId,
          type: 'CREDIT',
          amount,
          balanceBefore: availableBefore,
          balanceAfter: updatedWallet.availableBalance,
          description: 'Session completed - balance released',
          referenceType: metadata?.referenceType || 'session',
          referenceId: metadata?.referenceId,
          metadata: {
            ...metadata,
            pendingBefore,
            pendingAfter: updatedWallet.pendingBalance,
          },
        },
      });

      logger.info('Released pending balance', {
        providerId,
        amount,
        availableBalance: updatedWallet.availableBalance,
      });

      return updatedWallet;
    });
  }

  /**
   * Request payout
   * Provider initiates withdrawal
   */
  async requestPayout(providerId: string, amount: number) {
    // Validate minimum amount
    if (amount < env.MIN_PAYOUT_AMOUNT) {
      throw new AppError(
        `Minimum payout amount is ₹${env.MIN_PAYOUT_AMOUNT / 100}`,
        HTTP_STATUS.BAD_REQUEST,
        ERROR_CODES.PAYOUT_BELOW_MINIMUM
      );
    }

    return await prisma.$transaction(async (tx) => {
      // Lock wallet
      const wallet = await tx.providerWallet.findUnique({
        where: { providerId },
      });

      if (!wallet) {
        throw new AppError('Wallet not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
      }

      await tx.$queryRaw`
        SELECT * FROM provider_wallets 
        WHERE provider_id = ${providerId}
        FOR UPDATE
      `;

      // Check sufficient balance
      if (wallet.availableBalance < amount) {
        throw new AppError(
          'Insufficient available balance',
          HTTP_STATUS.BAD_REQUEST,
          ERROR_CODES.INSUFFICIENT_BALANCE
        );
      }

      // Create payout request
      const payoutRequest = await tx.payoutRequest.create({
        data: {
          walletId: wallet.id,
          userId: providerId,
          amount,
          status: PayoutStatus.REQUESTED,
        },
      });

      logger.info('Payout requested', {
        providerId,
        amount,
        payoutRequestId: payoutRequest.id,
      });

      return payoutRequest;
    });
  }

  /**
   * Approve payout (Admin only)
   * 
   * CRITICAL: Uses transaction with row-level locking
   */
  async approvePayout(adminId: string, payoutRequestId: string, transactionReference?: string) {
    return await prisma.$transaction(async (tx) => {
      // Lock payout request
      const payoutRequest = await tx.payoutRequest.findUnique({
        where: { id: payoutRequestId },
        include: { wallet: true },
      });

      if (!payoutRequest) {
        throw new AppError('Payout request not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
      }

      // Check status
      if (payoutRequest.status !== PayoutStatus.REQUESTED) {
        throw new AppError(
          'Payout already processed',
          HTTP_STATUS.BAD_REQUEST,
          ERROR_CODES.PAYOUT_ALREADY_PROCESSED
        );
      }

      // Lock wallet
      await tx.$queryRaw`
        SELECT * FROM provider_wallets 
        WHERE id = ${payoutRequest.walletId}
        FOR UPDATE
      `;

      const wallet = payoutRequest.wallet;

      // Check sufficient balance
      if (wallet.availableBalance < payoutRequest.amount) {
        throw new AppError(
          'Insufficient available balance',
          HTTP_STATUS.BAD_REQUEST,
          ERROR_CODES.INSUFFICIENT_BALANCE
        );
      }

      const balanceBefore = wallet.availableBalance;
      const processingFee = Math.floor(payoutRequest.amount * 0.02); // 2% processing fee
      const netAmount = payoutRequest.amount - processingFee;

      // Update wallet
      const updatedWallet = await tx.providerWallet.update({
        where: { id: wallet.id },
        data: {
          availableBalance: { decrement: payoutRequest.amount },
          totalWithdrawn: { increment: netAmount },
        },
      });

      // Update payout request
      const updatedPayout = await tx.payoutRequest.update({
        where: { id: payoutRequestId },
        data: {
          status: PayoutStatus.APPROVED,
          approvedBy: adminId,
          approvedAt: new Date(),
          processingFee,
          netAmount,
          transactionReference,
        },
      });

      // Record wallet transaction
      await tx.walletTransaction.create({
        data: {
          walletId: wallet.id,
          userId: payoutRequest.userId,
          type: 'DEBIT',
          amount: payoutRequest.amount,
          balanceBefore,
          balanceAfter: updatedWallet.availableBalance,
          description: `Payout approved - ${transactionReference || 'manual transfer'}`,
          referenceType: 'payout',
          referenceId: payoutRequestId,
          metadata: {
            processingFee,
            netAmount,
            approvedBy: adminId,
          },
        },
      });

      // Audit log
      await tx.auditLog.create({
        data: {
          userId: adminId,
          action: 'APPROVE_PAYOUT',
          entityType: 'payout_request',
          entityId: payoutRequestId,
          changes: {
            amount: payoutRequest.amount,
            providerId: payoutRequest.userId,
            transactionReference,
          },
        },
      });

      logger.info('Payout approved', {
        payoutRequestId,
        providerId: payoutRequest.userId,
        amount: payoutRequest.amount,
        netAmount,
        approvedBy: adminId,
      });

      return updatedPayout;
    });
  }

  /**
   * Reject payout (Admin only)
   */
  async rejectPayout(adminId: string, payoutRequestId: string, reason: string) {
    const payoutRequest = await prisma.payoutRequest.findUnique({
      where: { id: payoutRequestId },
    });

    if (!payoutRequest) {
      throw new AppError('Payout request not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
    }

    if (payoutRequest.status !== PayoutStatus.REQUESTED) {
      throw new AppError(
        'Payout already processed',
        HTTP_STATUS.BAD_REQUEST,
        ERROR_CODES.PAYOUT_ALREADY_PROCESSED
      );
    }

    const updatedPayout = await prisma.payoutRequest.update({
      where: { id: payoutRequestId },
      data: {
        status: PayoutStatus.REJECTED,
        approvedBy: adminId,
        approvedAt: new Date(),
        rejectionReason: reason,
      },
    });

    // Audit log
    await prisma.auditLog.create({
      data: {
        userId: adminId,
        action: 'REJECT_PAYOUT',
        entityType: 'payout_request',
        entityId: payoutRequestId,
        changes: {
          amount: payoutRequest.amount,
          providerId: payoutRequest.userId,
          reason,
        },
      },
    });

    logger.info('Payout rejected', {
      payoutRequestId,
      providerId: payoutRequest.userId,
      reason,
      rejectedBy: adminId,
    });

    return updatedPayout;
  }

  /**
   * Get wallet by provider ID
   */
  async getWallet(providerId: string) {
    const wallet = await prisma.providerWallet.findUnique({
      where: { providerId },
      include: {
        transactions: {
          orderBy: { createdAt: 'desc' },
          take: 10,
        },
        payouts: {
          orderBy: { createdAt: 'desc' },
          take: 10,
        },
      },
    });

    if (!wallet) {
      throw new AppError('Wallet not found', HTTP_STATUS.NOT_FOUND, ERROR_CODES.RESOURCE_NOT_FOUND);
    }

    return wallet;
  }
}

File: backend/wallet_service.py
import os
import uuid
from datetime import datetime, timezone

from database import (
    provider_wallets_col, wallet_transactions_col, payout_requests_col
)

COMMISSION_PERCENT = int(os.environ.get("PLATFORM_COMMISSION_PERCENT", "40"))
THERAPIST_SHARE = int(os.environ.get("THERAPIST_SHARE_PERCENT", "60"))
MIN_PAYOUT = int(os.environ.get("MIN_PAYOUT_AMOUNT", "100000"))


def calculate_split(total_amount: int) -> dict:
    provider_earning = int(total_amount * THERAPIST_SHARE / 100)
    platform_commission = total_amount - provider_earning
    return {
        "provider_earning": provider_earning,
        "platform_commission": platform_commission,
    }


async def get_wallet(provider_id: str) -> dict:
    wallet = await provider_wallets_col.find_one(
        {"provider_id": provider_id}, {"_id": 0}
    )
    if not wallet:
        raise ValueError("Wallet not found")
    return wallet


async def credit_pending(provider_id: str, amount: int, description: str, ref_type: str = None, ref_id: str = None):
    wallet = await get_wallet(provider_id)
    now = datetime.now(timezone.utc).isoformat()
    balance_before = wallet["pending_balance"]
    balance_after = balance_before + amount

    await provider_wallets_col.update_one(
        {"provider_id": provider_id},
        {"$inc": {"pending_balance": amount}, "$set": {"updated_at": now}}
    )

    tx = {
        "id": str(uuid.uuid4()),
        "wallet_id": wallet["id"],
        "user_id": provider_id,
        "type": "CREDIT",
        "amount": amount,
        "balance_before": balance_before,
        "balance_after": balance_after,
        "description": description,
        "reference_type": ref_type,
        "reference_id": ref_id,
        "created_at": now,
    }
    await wallet_transactions_col.insert_one(tx)
    return tx


async def settle_pending(provider_id: str, amount: int, description: str, ref_type: str = None, ref_id: str = None):
    wallet = await get_wallet(provider_id)
    if wallet["pending_balance"] < amount:
        raise ValueError("Insufficient pending balance")

    now = datetime.now(timezone.utc).isoformat()

    await provider_wallets_col.update_one(
        {"provider_id": provider_id},
        {
            "$inc": {
                "pending_balance": -amount,
                "available_balance": amount,
                "lifetime_earnings": amount,
            },
            "$set": {"updated_at": now}
        }
    )

    tx = {
        "id": str(uuid.uuid4()),
        "wallet_id": wallet["id"],
        "user_id": provider_id,
        "type": "CREDIT",
        "amount": amount,
        "balance_before": wallet["available_balance"],
        "balance_after": wallet["available_balance"] + amount,
        "description": description,
        "reference_type": ref_type,
        "reference_id": ref_id,
        "created_at": now,
    }
    await wallet_transactions_col.insert_one(tx)


async def request_payout(provider_id: str, amount: int) -> dict:
    wallet = await get_wallet(provider_id)
    if wallet["available_balance"] < amount:
        raise ValueError("Insufficient available balance")
    if amount < MIN_PAYOUT:
        raise ValueError(f"Minimum payout amount is {MIN_PAYOUT} paise")

    now = datetime.now(timezone.utc).isoformat()
    payout = {
        "id": str(uuid.uuid4()),
        "wallet_id": wallet["id"],
        "provider_id": provider_id,
        "amount": amount,
        "status": "REQUESTED",
        "created_at": now,
        "updated_at": now,
    }
    await payout_requests_col.insert_one(payout)

    # Deduct from available balance
    await provider_wallets_col.update_one(
        {"provider_id": provider_id},
        {"$inc": {"available_balance": -amount}, "$set": {"updated_at": now}}
    )

    tx = {
        "id": str(uuid.uuid4()),
        "wallet_id": wallet["id"],
        "user_id": provider_id,
        "type": "DEBIT",
        "amount": amount,
        "balance_before": wallet["available_balance"],
        "balance_after": wallet["available_balance"] - amount,
        "description": f"Payout request #{payout['id'][:8]}",
        "reference_type": "payout",
        "reference_id": payout["id"],
        "created_at": now,
    }
    await wallet_transactions_col.insert_one(tx)

    payout.pop("_id", None)
    return payout


async def process_payout_action(payout_id: str, action: str, admin_id: str, reason: str = None) -> dict:
    payout = await payout_requests_col.find_one({"id": payout_id}, {"_id": 0})
    if not payout:
        raise ValueError("Payout request not found")
    if payout["status"] != "REQUESTED":
        raise ValueError("Payout already processed")

    now = datetime.now(timezone.utc).isoformat()

    if action == "approve":
        await payout_requests_col.update_one(
            {"id": payout_id},
            {"$set": {
                "status": "APPROVED",
                "approved_by": admin_id,
                "approved_at": now,
                "updated_at": now,
            }}
        )
        await provider_wallets_col.update_one(
            {"provider_id": payout["provider_id"]},
            {"$inc": {"total_withdrawn": payout["amount"]}, "$set": {"last_payout_at": now, "updated_at": now}}
        )
    elif action == "reject":
        await payout_requests_col.update_one(
            {"id": payout_id},
            {"$set": {
                "status": "REJECTED",
                "rejection_reason": reason,
                "approved_by": admin_id,
                "updated_at": now,
            }}
        )
        # Refund balance
        await provider_wallets_col.update_one(
            {"provider_id": payout["provider_id"]},
            {"$inc": {"available_balance": payout["amount"]}, "$set": {"updated_at": now}}
        )

    updated = await payout_requests_col.find_one({"id": payout_id}, {"_id": 0})
    return updated
